homeassistant/components/home_plus_control/api.py

```diff
@@ -0,0 +1,341 @@
+            except KeyError:
+                new_interval_value = DEFAULT_UPDATE_INTERVALS[interval_name]
+
+            self._refresh_intervals[interval_name] = new_interval_value
+            self.logger.debug(
+                "Updated API refresh interval: %s: %s",
+                interval_name,
+                self._refresh_intervals[interval_name],
+            )
+
+    async def async_handle_plant_data(self):
+        """Recover the plant data for this particular user.
+
+        This will populate the "private" array of plants of this object and will return it.
+        It is expected that in most cases, there will only be one plant.
+
+        Returns:
+            dict: Dictionary of plants for this user - Keyed by the plant ID. Can be empty if no
+                  plants are retrieved.
+        """
+        # Attempt to recover the plant information from the Hass data object.
+        # If it is not there, then we request it from the API and add it.
+        # We also refresh from the API if the time has expired.
+        plant_info = self.hass.data[self._domain].get("plant_info")
+        if plant_info is None or self._should_check(
+            CONF_PLANT_UPDATE_INTERVAL,
+            self._refresh_intervals[CONF_PLANT_UPDATE_INTERVAL],
+        ):
+            try:
+                result = await self.get_request(PLANT_URL)  # Call the API
+                plant_info = await result.json()
+            except aiohttp.ClientError as err:
+                raise HomeAssistantError("Error retrieving plant information") from err
+
+            # If all goes well, we update the last check time
+            self._last_check[CONF_PLANT_UPDATE_INTERVAL] = time.monotonic()
+
+            self.hass.data[self._domain]["plant_info"] = plant_info
+            self.logger.debug("Obtained plant information from API: %s", plant_info)
+        else:
+            self.logger.debug(
+                "Obtained plant information from Hass object: %s", plant_info
+            )
+
+        # Populate the dictionary of plants
+        current_plant_ids = []
+        for plant in plant_info["plants"]:
+            current_plant_ids.append(plant["id"])
+            if plant["id"] in self._plants:
+                self.logger.debug(
+                    "Plant with id %s is already cached. Only updating the existing data.",
+                    plant["id"],
+                )
+                cur_plant = self._plants.get(plant["id"])
+                # We will update the plant info just in case and ensure it has an Api object
+                cur_plant.name = plant["name"]
+                cur_plant.country = plant["country"]
+                if cur_plant.oauth_client is None:
+                    cur_plant.oauth_client = self
+            else:
+                self.logger.debug("New plant with id %s detected.", plant["id"])
+                self._plants[plant["id"]] = HomePlusPlant(
```

Comment:
This should be handled by the library.

Looking at this class, most things should just be handled by the library and we shouldn't need to do any management of the plants or modules in Home Assistant. The only reason why we're implementing a subclass of the parent class is to be able to handle token refresh. That's the only part that we should need to do here. Move everything else to the library.

In the coordinator we should be able to retrieve new data and compare with the existing data to see if we need to remove entities.

Reply:
OK, let me work then on making the changes here and in the library :+1:

Reply:
I extracted all of this functionality to the library. The API class looks much cleaner now :-) and it simply recovers the latest status of the modules. The library requirement has been upgraded to v0.0.4.